### PR TITLE
XRANGE support exclusive range

### LINKED ISSUE #2631: please implements XRANGE ( Exclusive ranges param !
please implements XRANGE ( Exclusive ranges param !

### PR DESCRIPTION
**Pull request by AvitalFineRedis**
Closes #2631

### CODE REVIEW

File: src/main/java/redis/clients/jedis/Client.java [hunk 1]
```diff
@@ -1507,6 +1507,12 @@ public void xrange(final String key, final StreamEntryID start, final StreamEntr
       SafeEncoder.encode(end == null ? "+" : end.toString()), count);
   }
 
+  @Override
+  public void xrange(String key, String start, String end, int count) {
```
Comment (yangbodong22011) [inline src/main/java/redis/clients/jedis/Client.java:1511]: we should also support binary commands?
```
public void xrange(byte[] key, byte[] start, byte[] end, int count) {
```

Comment (sazzad16) [reply src/main/java/redis/clients/jedis/Client.java:1511]: @yangbodong22011 those are already available in `Binary...` classes.

Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/Client.java:1511]: @sazzad16 ohh...yes, ignore my comment

**Comment by sazzad16** [comment]
Please check that this PR, as of now, is a breaking change. I.e. any invocation of 

    jedis.xrange(streamName, null, null, count);

must be changed to any of following (or something else)

    xrange(streamName, (StreamEntryID) null, (StreamEntryID) null, count);
    xrange(streamName, (String) null, (String) null, count);

I can think of two different options

1. target this PR for 4.0.0 and mention the breaking behavior in release notes
2. rename methods (e.g. xrange2/xrevrange2 or xrangeV2/xrevrangeV2)

Please share your thoughts.

@yangbodong22011 @dengliming @mina-asham

**Comment by yangbodong22011** [comment]
> I can think of two different options
> 
> 1. target this PR for 4.0.0 and mention the breaking behavior in release notes
> 2. rename methods (e.g. xrange2/xrevrange2 or xrangeV2/xrevrangeV2)

vote for 1，rename is not a good way, because the binary `xrange` already exists and may be used by the user.
```
public void xrange(final byte[] key, final byte[] start, final byte[] end, final int count);
```

**Comment by AvitalFineRedis** [comment]
> Please check that this PR, as of now, is a breaking change. I.e. any invocation of
> 
> ```
> jedis.xrange(streamName, null, null, count);
> ```
> 
> must be changed to any of following (or something else)
> 
> ```
> xrange(streamName, (StreamEntryID) null, (StreamEntryID) null, count);
> xrange(streamName, (String) null, (String) null, count);
> ```
> 
> I can think of two different options
> 
> 1. target this PR for 4.0.0 and mention the breaking behavior in release notes
> 2. rename methods (e.g. xrange2/xrevrange2 or xrangeV2/xrevrangeV2)
> 
> Please share your thoughts.
> 
> @yangbodong22011 @dengliming @mina-asham

vote for 1 also

**Comment by AvitalFineRedis** [comment]
Vote for 1 also

**Comment by dengliming** [comment]
> I can think of two different options
> 
> 1. target this PR for 4.0.0 and mention the breaking behavior in release notes
> 2. rename methods (e.g. xrange2/xrevrange2 or xrangeV2/xrevrangeV2)

Vote for 1, agree with @yangbodong22011

**Comment by sazzad16** [comment]
@yangbodong22011 @dengliming We are done voting. I have set this for 4.0.0. Please review the code 🙂

**Comment by dengliming** [comment]
Actually, I'm thinking about introducing class `Range` ( refer to [https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/Range.java](https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/Range.java), and the method name like the following 

`jedis.xrange(streamName, range, count);`

Would it be better?

**Comment by sazzad16** [comment]
@dengliming It'd be great! 

But this would bring the feature early and it won't break your plan either.

**Pull request closed without merging**